[export] Add stack_trace for non-strict export

pianpwk opened the pull request:
This addresses 2 issues with stack_trace metadata:
- stack_trace is currently missing from nodes in non-strict export
- in strict mode, stack_trace is populated for placeholder nodes, which may not be well-defined (with multiple uses)

We filter the call stack during tracing for calls from forward() methods, or ops in `torch.__init__.py` (e.g. sym_size_int, sym_constrain_range, etc.) to populate stack_trace. A node-level check is also added to _export_non_strict().

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

angelayi commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -590,6 +591,35 @@ def unwrap_proxy(self, e):
         else:
             return e
 
+    def create_node(self, *args, **kwargs):
```
Maybe we can add to ModuleStackTracer [here](https://fburl.com/code/2hh7kp7a) instead of to the PythonKeyTracer? Just in case normal make_fx users do not care ab having metadata.

angelayi commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -590,6 +591,35 @@ def unwrap_proxy(self, e):
+        if 'stack_trace' not in node.meta and node.op not in ["placeholder", "output"]:
+            user_frame = self._find_user_frame()
+            if user_frame:
+                summary = traceback.extract_stack(user_frame)
```
I wonder if maybe we could try out [`CapturedTraceback`](https://github.com/pytorch/pytorch/pull/107358?fbclid=IwAR1u_DAexSNr1ASYLsRUJ0jS7xPiqSjSKlT2-txdkXRUPF-WD5HSaFXxYzw)? It seems to be a faster version of `traceback.extract_stack`

angelayi commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -590,6 +591,35 @@ def unwrap_proxy(self, e):
+                # this is hardcoded, but leads to a much cleaner stack trace
+                stack_trace = [
+                    frame for frame in stack_trace if not (
+                        frame.filename.endswith('fx/_symbolic_trace.py')
```
I'm not super familiar with what `_find_user_frame` does, but it seems like it already filters out `fx/_symbolic_trace.py` [here](https://fburl.com/code/ytmmswlk)? 

Separately -- maybe we could rewrite this filtering to match the code in `_find_user_frame`, where the files we want to filter are in a list rather than having multiple `frame.filename.endswith` calls

pianpwk replied:
Thank you for the comments! Just addressed these.

For `_find_user_frame` I think it only filters for the one returned frame, but the stack might still contain those files

angelayi reviewed (commented):
Looks good! Do you mind adding a test to test/export/test_export.py so that we can track regressions to the stack trace metadata?

pianpwk committed:
[export] Add stack_trace for non-strict export (#121034)

Summary:
This addresses 2 issues with stack_trace metadata:
- stack_trace is currently missing from nodes in non-strict export
- in strict mode, stack_trace is populated for placeholder nodes, which may not be well-defined (with multiple uses)

We filter the call stack during tracing for calls from forward() methods, or ops in `torch.__init__.py` (e.g. sym_size_int, sym_constrain_range, etc.) to populate stack_trace. A node-level check is also added to _export_non_strict().


Differential Revision: D54917080

Pulled By: pianpwk

pianpwk committed:
merge conflict

pianpwk committed:
test case

pianpwk committed:
stack trace with traceback

pianpwk committed:
Merge branch 'main' of https://github.com/pytorch/pytorch into pianpwk/nonstrict_stack_trace

pianpwk committed:
wrong changes, can't export from internal

angelayi reviewed (approved):
nice! 😎
